pkg/registry/apis/provisioning/jobs/changes.go

```diff
@@ -0,0 +1,89 @@
+package jobs
+
+import (
+	"fmt"
+	"sort"
+	"strings"
+
+	provisioning "github.com/grafana/grafana/pkg/apis/provisioning/v0alpha1"
+	"github.com/grafana/grafana/pkg/registry/apis/provisioning/repository"
+	"github.com/grafana/grafana/pkg/registry/apis/provisioning/resources"
+)
+
+type ResourceFileChange struct {
+	// Path to the file in a repository with a change
+	Path   string
+	Action repository.FileAction
+
+	// The current value in the database -- only required for delete
+	Existing *provisioning.ResourceListItem
+}
+
+func Changes(source []repository.FileTreeEntry, target *provisioning.ResourceList) ([]ResourceFileChange, error) {
+	lookup := make(map[string]*provisioning.ResourceListItem, len(target.Items))
+	for _, item := range target.Items {
+		if item.Path == "" {
+			if item.Group != "folder.grafana.app" {
+				return nil, fmt.Errorf("empty path on a non folder")
+			}
+			continue
+		}
+		lookup[item.Path] = &item
+	}
+
+	var keep []string
+	changes := make([]ResourceFileChange, 0, len(source))
+	for _, file := range source {
+		if !file.Blob {
+			continue // skip folder references?
+		}
+
+		check, ok := lookup[file.Path]
+		if ok {
+			if check.Hash != file.Hash {
```

Comment:
[style] let's use guard clauses when possible to reduce the nested control blocks and indentations: 

https://refactoring.guru/replace-nested-conditional-with-guard-clauses

Reply:
maybe better 😬  -- this has grown quite a bit so the diff is really hard to compare!

Reply:
there is a bit of room for improvement in this function in terms of naming and those control structures but overall it looks much better now with only comparison logic and not execution. 

Great job!